doc/source/conf.py

```diff
@@ -545,6 +544,11 @@ def linkcode_resolve(domain, info):
     if isinstance(obj, type) and obj.__module__ == 'numpy':
         fn = _get_c_source_file(obj)
 
+    # This can be removed when removing the decorator set_module. Fix issue #28629
```

Comment:
I think this can logically be folded into the previous `if` statement, since `_module_file` can only be set if that statement holds. So maybe,
```
    if isinstance(obj, type) and obj.__module__ == 'numpy':
        # This can be removed when removing the decorator set_module. Fix issue #28629
        if (fn := getattr(obj, '_module_file', None) is not None:
            fn = relpath(fn, start=dirname(numpy.__file__))
        else:
            fn = _get_c_source_file(obj)

if fn is None:
    ...
```

I guess the above makes me wonder too whether we can set `_module_file` for C source code... But for another time!